proposals/NNNN-isolated-any-functions.md

```diff
@@ -0,0 +1,671 @@
+# `@isolated(any)` Function Types
+
+* Proposal: [SE-NNNN](NNNN-isolated-any-functions.md)
+* Authors: [John McCall](https://github.com/rjmccall)
+* Review Manager: TBD
+* Status: **Awaiting implementation**
+* Implementation: [apple/swift#71433](https://github.com/apple/swift/pull/71433), [apple/swift#71574](https://github.com/apple/swift/pull/71574)
+* Review: ([pitch](https://forums.swift.org/t/isolated-any-function-types/70562))
+
+[SE-0316]: https://github.com/apple/swift-evolution/blob/main/proposals/0316-global-actors.md
+[SE-0392]: https://github.com/apple/swift-evolution/blob/main/proposals/0392-custom-actor-executors.md
+[isolated-captures]: https://forums.swift.org/t/closure-isolation-control/70378
+
+## Introduction
+
+The actor isolation of a function is an important part of how it's
+used.  Swift can reason precisely about the isolation of a specific
+function *declaration*, but when functions are passed around as
+*values*, Swift's function types are not expressive enough to keep up.
+This proposal adds a new kind of function type that carries its function's
+actor isolation dynamically.  This solves a variety of expressivity
+problems in the language.  It also allows features such as the standard
+library's task-creation APIs to be implemented more efficiently and
+with stronger semantic guarantees.
+
+## Motivation
+
+The safety of Swift concurrency relies on understanding the isolation
+requirements of functions.  The caller of an isolated synchronous
+function must run it in an appropriately-isolated context or else the
+function will almost certainly introduce data races.
+
+Function declarations and closures in Swift support three different
+forms of actor isolation:
+
+- They can be non-isolated.
+
+  - This can be explicitly expressed with the `nonisolated` modifier.
+  - It is also the default if other rules don't apply.
+
+- They can be isolated to a specific [global actor][SE-0316] type.
+
+  - This can be explicitly expressed with a global actor attribute,
+    such as `@MainActor`.
+  - It can also be inferred from context in a number of ways, such as
+    if the function is a method of a type with that attribute (which
+    itself can be inferred in a number of ways).
+  - Additionally, a closure passed directly to the `Task` initializer
```

Comment:
A small drafting suggestion: the bullets here are interspersing the kinds of isolation with the ways in which isolation is inferred, making it harder to keep both separate in the reader's mind. Perhaps lay out the kinds of isolation here, and move the "inference" bits until later?